xmake/rules/c++/modules/modules_support/builder.lua

```diff
@@ -0,0 +1,407 @@
+--!A cross-platform build utility based on Lua
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+-- Copyright (C) 2015-present, TBOOX Open Source Group.
+--
+-- @author      ruki, Arthapz
+-- @file        common.lua
+--
+
+-- imports
+import("core.base.json")
+import("core.base.option")
+import("private.async.buildjobs")
+import("core.tool.compiler")
+import("core.project.config")
+import("core.project.depend")
+import("utils.progress")
+import("compiler_support")
+import("dependency_scanner")
+
+-- build module files
+function _build_modules(target, batch, objectfiles, modules, opt)
+
+    -- batchcmd stuff
+    local depmtime = 0
+
+    -- batchjobs stuff
+    local modulesjobs = {}
+    local tailjob = _builder(target).init_build_for(target, batch, modules, table.join({type = "module"}, opt))
+    table.join2(modulesjobs, tailjob)
+
+    -- flush job
+    if opt.batchjobs then 
+        opt.rootjob = batch:addjob(target:name() .. "_flush_mapper", function(index, total)
+            progress.show((index * 100) / total, "${color.build.object}flushing.module.map %s", target:name())
+            _flush_target_mapper(target)
+        end, {rootjob = opt.rootjob})
+    end
+
+    -- build modules
+    for _, objectfile in ipairs(objectfiles) do
+        local module = modules[objectfile]
+        if module then
+            local name, provide, cppfile = get_provided_module(module)
+            cppfile = cppfile or module.cppfile
+
+            local fileconfig = target:fileconfig(cppfile)
+
+            -- add objectfile if target is not static lib and is named module or if it's an implementation module
+            if target:kind() ~= "static" and provide or compiler_support.has_module_extension(cppfile) then
+                target:add("objectfiles", objectfile)
+            end
+
+            local bmifile = provide and compiler_support.get_bmi_path(provide.bmi)
+            local build = _should_build(target, cppfile, bmifile, objectfile, module.requires)
+            if not compiler_support.memcache():get2(target:name() .. cppfile, "compiling") then
+                compiler_support.memcache():set2(target:name() .. cppfile, "compiling", build)
+                -- for rebuild
+                if name then
+                    compiler_support.memcache():set2(name, "compiling", build)
+                end
+
+                if opt.batchjobs then
```

Comment:
I think it would be clearer and better maintained to have two separate functions for batchjob and batchcmds.

_build_modules_for_batchjobs
_build_modules_for_batchcmds